### PR TITLE
[SYM-4767] Runtime Connector Module Version 2 Upgrade Guide

### PR DESCRIPTION
**Pull request by llam15**
# Summary
- Adds an upgrade guide for Version 2 under `docs/`
- Links to this upgrade guide from the main `README.md`
- After more careful reading of the [Terraform refactoring docs](https://developer.hashicorp.com/terraform/language/modules/develop/refactoring#moved-block-syntax) and this [GitHub Issue](https://github.com/hashicorp/terraform/issues/30182), adds a `v2.0.0-moved.tf` to handle the renaming of the `aws_iam_role.this` and `aws_iam_role_policy_attachment.assume_roles_attach`, both of which are module internal resources whose renaming should be the module's responsibility.

### CODE REVIEW

**Comment by llam15** [comment]
Current dependencies on/for this PR:
* main
  * **PR #7** <a href="https://app.graphite.dev/github/pr/symopsio/terraform-aws-runtime-connector/7" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/symopsio/terraform-aws-runtime-connector/7?utm_source=stack-comment).

File: docs/version-2-upgrade.md [hunk 1]
```diff
@@ -0,0 +1,385 @@
+---
+subcategory: ""
+page_title: "Runtime Connector Module Version 2 Upgrade Guide"
+description: |-
+  Runtime Connector Module Version 2 Upgrade Guide
+---
+
+# Runtime Connector Module Version 2 Upgrade Guide
+
+Version 2.0.0 of the Runtime Connector Module a major release and includes some changes that you will need to consider when upgrading. This guide is intended to help with that process.
```
Comment (sym-andrewa) [inline docs/version-2-upgrade.md]:
```suggestion
Version 2.0.0 of the Runtime Connector Module is a major release and includes some changes that you will need to consider when upgrading. This guide is intended to help with that process.
```
nit: grammar

File: docs/version-2-upgrade.md [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+```terraform
+module "runtime_connector" {
+  source  = "symopsio/runtime-connector/aws"
+  version = "~> 2.0"
+
+  environment = var.environment
+}
+```
+
+Optionally, you may choose to use `moved` configuration blocks to migrate your Terraform state, instead of replacing the existing AWS IAM Role and Policy with a new one:
```
Comment (sym-andrewa) [inline docs/version-2-upgrade.md]:
```suggestion
Optionally, you may also use `moved` configuration blocks to migrate your Terraform state, instead of replacing the existing AWS IAM Role and Policy with a new one:
```
Nit: Clarify that this is something you do in addition to the version change, not instead of the version change. Might also be good to link to the Terraform docs that explain how `moved` blocks work?

File: docs/version-2-upgrade.md [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+module "secrets_manager_access" {
+  source  = "symopsio/secretsmgr-addon/aws"
+  version = "~> 1.1"
+
+  environment   = var.environment
+  iam_role_name = module.runtime_connector.sym_runtime_connector_role.name
+}
+```
+
+Optionally, you may choose to use `moved` configuration blocks to migrate your Terraform state, instead of replacing the existing IAM policies with new ones:
```
Comment (sym-andrewa) [inline docs/version-2-upgrade.md]:
```suggestion
Optionally, you may also use `moved` configuration blocks to migrate your Terraform state, instead of replacing the existing IAM policies with new ones:
```

File: docs/version-2-upgrade.md [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+module "kinesis_firehose_access" {
+  source  = "symopsio/kinesis-firehose-addon/aws"
+  version = "~> 1.1"
+
+  environment = var.environment
+  iam_role_name = module.runtime_connector.sym_runtime_connector_role.name
+}
+```
+
+Optionally, you may choose to use `moved` configuration blocks to migrate your Terraform state, instead of replacing the existing IAM policies with new ones:
```
Comment (sym-andrewa) [inline docs/version-2-upgrade.md]:
```suggestion
Optionally, you may also use `moved` configuration blocks to migrate your Terraform state, instead of replacing the existing IAM policies with new ones:
```

File: docs/version-2-upgrade.md [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+  source  = "symopsio/kinesis-data-stream-addon/aws"
+  version = "~> 1.1"
+
+  environment   = var.environment
+  stream_arns   = ["arn:aws:kinesis:*:111122223333:stream/my-stream"]
+  iam_role_name = module.runtime_connector.sym_runtime_connector_role.name
+}
+```
+
+Optionally, you may choose to use `moved` configuration blocks to migrate your Terraform state, instead of replacing the existing IAM policies with new ones:
```
Comment (sym-andrewa) [inline docs/version-2-upgrade.md]:
```suggestion
Optionally, you may also use `moved` configuration blocks to migrate your Terraform state, instead of replacing the existing IAM policies with new ones:
```

File: docs/version-2-upgrade.md [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+  from = module.runtime_connector.aws_iam_role_policy_attachment.aws_kinesis_data_stream_attach[0]
+  to   = module.kinesis_data_stream_access.aws_iam_role_policy_attachment.attach_datastream_access[0]  
+}
+```
+
+## Removed Input: `custom_external_id`
+This input is no longer supported, and the `external_id` of the AWS IAM Role declared by this module will always be a random UUID.
+
+## Input `policy_arns` has been removed
+This input is no longer supported, and any additional AWS IAM Policies should be attached directly to the AWS IAM Role outputted by the module.
```
Comment (sym-andrewa) [inline docs/version-2-upgrade.md]:
```suggestion
This input is no longer supported, and any additional AWS IAM Policies should be attached directly to the AWS IAM Role output by the module.
```
"Outputted" has always read weird to me; though it is technically correct, I think using "output" is more common.

File: docs/version-2-upgrade.md [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+  type = "aws_secrets_manager"
+  name = "${var.environment}-secrets"
+
+  settings = {
+    context_id = module.runtime_connector.sym_integration.id
+  }
+}
+```
+
+Optionally, you may choose to use `moved` configuration blocks to migrate your Terraform state, instead of destroying and recreating your existing `sym_integration.runtime_context`:
```
Comment (sym-andrewa) [inline docs/version-2-upgrade.md]:
```suggestion
Optionally, you may also use `moved` configuration blocks to migrate your Terraform state, instead of destroying and recreating your existing `sym_integration.runtime_context`:
```

File: docs/version-2-upgrade.md [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+  environment = var.environment  
+}
+
+resource "sym_environment" "this" {
+  name       = var.environment
+  runtime_id = module.runtime_connector.sym_runtime.id
+}
+```
+
+Optionally, you may choose to use `moved` configuration blocks to migrate your Terraform state, instead of destroying and recreating your existing `sym_runtime`:
```
Comment (sym-andrewa) [inline docs/version-2-upgrade.md]:
```suggestion
Optionally, you may also use `moved` configuration blocks to migrate your Terraform state, instead of destroying and recreating your existing `sym_runtime`:
```

**Review by sym-andrewa — APPROVED** [review]
LGTM, just some nits! Just to confirm though—a lot of the migrations discussed here use phrasing along the lines of, "you may choose to use this integration instead of...," the implication here being that this is an optional migration and nothing will break if you keep it as it is currently. Is that correct?

File: docs/version-2-upgrade.md (hunk 6 above)
Comment (llam15) [reply docs/version-2-upgrade.md]: oh good it wasn't just me; I waffled back and forth between `output` and `outputted` when writing this 😂

**Comment by llam15** [comment]
> the implication here being that this is an optional migration and nothing will break if you keep it as it is currently. Is that correct?

Hmm good call out. Theoretically there is nothing wrong with declaring the `sym_integration.runtime_context` and the `sym_runtime.this` explicitly. Though if you don't remove the `sym_integration.runtime_context` and use the one built into the module, you will need to refactor the settings because the module no longer has a `settings` output.

I'll reword this to recommend using the `sym_integration` output by the module, since the `permission_context` settings a pretty confusing to declare explicitly.
